content/references/rippled-api/api-conventions/serialization.md

```diff
@@ -0,0 +1,285 @@
+| `FIELDS`              | A sorted array of tuples representing all fields that may appear in transactions, ledger objects, or other data. The first member of each tuple is the string name of the field and the second member is an object with that field's properties. (See the "Field properties" table below for definitions of those fields.) |
+| `TRANSACTION_RESULTS` | Map of [transaction result codes](transaction-results.html) to their numeric values. Result types not included in ledgers have negative values;`tesSUCCESS` has numeric value 0; [`tec`-class codes](tec-codes.html) represent failures that are included in ledgers. |
+| `TRANSACTION_TYPES`   | Map of all [transaction types](transaction-types.html) to their numeric values. |
+
+For purposes of serializing transactions for signing and submitting, the `FIELDS`, `TYPES`, and `TRANSACTION_TYPES` fields are necessary.
+
+The field definition objects in the `FIELDS` array have the following fields:
+
+| Field            | Type    | Contents                                        |
+|:-----------------|:--------|:------------------------------------------------|
+| `nth`            | Number  | The [field code](#field-codes) of this field, for use in constructing its [Field ID](#field-ids) and ordering it with other fields of the same data type. |
+| `isVLEncoded`    | Boolean | If `true`, this field is [variable-length encoded](#variable-length-encoding). |
+| `isSerialized`   | Boolean | If `true`, this field should be encoded into serialized binary data. When this field is `false`, the field is typically reconstructed on demand rather than stored. |
+| `isSigningField` | Boolean | If `true` this field should be serialized when preparing a transaction for signing. If `false`, this field should be omitted from the data to be signed. (It may not be part of transactions at all.) |
+| `type`           | String  | The internal data type of this field. This maps to a key in the `TYPES` map, which gives the [type code](#type-codes) for this field. |
+
+
+
+## Canonical Field Order
+
+All fields in a transaction are sorted in a specific order based on the field's type first, then the field itself second. (Think of it as sorting by family name, then given name, where the family name is the field's type and the given name is the field itself.)
+
+### Field IDs
+
+[[Source - Encoding]](https://github.com/seelabs/rippled/blob/cecc0ad75849a1d50cc573188ad301ca65519a5b/src/ripple/protocol/impl/Serializer.cpp#L117-L148 "Source")
+[[Source - Decoding]](https://github.com/seelabs/rippled/blob/cecc0ad75849a1d50cc573188ad301ca65519a5b/src/ripple/protocol/impl/Serializer.cpp#L484-L509 "Source")
+
+
+When you combine the type code and sort code, you get the field's identifier, which is prefixed before the field in the final serialized blob. The size of the Field ID is one to three bytes depending on the type code and field codes it combines. See the following table:
+
```

Comment:
Sorry if I'm reading too closely into this - but I'm hanging on every word to be able to grasp this info. 🤓 I'm a little confused by the difference between the field ID being a combination of the 

> type code and sort code

And then reading that the size of the Field ID depends on the 

> type code and field codes it combines

Is the field ID the combination of the type code and sort code, or the combination of the type code and field code? Or all three? I think I'm misunderstanding.

Reply:
Sorry, I made up too many terms and got them mixed up a bit. It's probably easiest to think of it as that there is no such thing as a "sort code". (I'll fix this.)

It's a little confusing, but basically:

- Sort by type code, then by field code.
- Identify a field by Field ID, which is the serialized combination of type code and field code.
- Don't actually sort by Field ID directly because the variable-byte-structure of it would screw up sorting.